### ISSUE TITLE
Unexplained rollbacks with PreSpawnedObjects

### ISSUE THREAD

**Issue by cBournhonesque** [cBournhonesque/lightyear#631]
Okay, going back to this rollback issue, I think I've narrowed it down enough to share: I might be running into an edge case bug with PreSpawnedPlayerObjects, probably.

To summarize: my game is using projectiles that are really fast moving. There projectiles are pre-spawned by the client that shoots. Whenever the projectiles are stopped, they get deleted that frame.
There's three types of behaviors I'm seeing with this, two of which are failure points:

- If the projectiles get destroyed really early (in the video example, when I shoot the wall I'm in front of) no rollbacks are triggered but an error is printed
```
ERROR lightyear::shared::replication::receive: Received despawn for an entity that does not exist
```

This error gets printed by both clients.
- If the projectiles get destroyed late enough, there's no errors and no rollbacks (this is the case where everything works correctly!)
- There's a middle ground between the first and second case where no errors are printed but a rollback is triggered in the client that fired the shot (this should not happen) on some of the projectile components (which component this is triggered for changes per game boot) due to a [missing history_value in the case where the confirmed entity exists](https://github.com/cBournhonesque/lightyear/blob/e9622df61ecaba8db005e9252d820e6330647a6b/lightyear/src/client/prediction/rollback.rs#L189)

See https://discord.com/channels/691052431525675048/1189344685546811564/1282437035604312187

**Comment by cBournhonesque** [comment]
- For `ERROR lightyear::shared::replication::receive: Received despawn for an entity that does not exist`. Not 100% sure what is happening. 
   - I think the entity is spawned/destroyed within a replication_interval, so the server just spawns a `Despawn` message for the entity but it didn't send a `Spawn` message. (`Spawn` messages are sent every `replication_interval` so if the entity was despawned before we run the system we wouldn't send any Spawn, but `Despawn` messages are buffered via observer whenever the `ReplicationTarget` component is removed, so we would **always** send a despawn no matter what, even if no `Spawned` was sent.
   - it's not a huge problem, but we'll have to think eventually about how to handle it. Maybe send a message with both `spawn` and `despawn` in the same message? Or just don't send the `despawn` at all?

**Comment by OlivierCoue** [comment]
Hi, I ran into the same issue. It seem to happen even when using the : `commands.entity(entity).remove::<server::Replicate>().despawn();` strat (that I took from your spaceships example to despawn without replication).

Same behavior, I get the error `ERROR lightyear::shared::replication::receive: Received despawn for an entity that does not exist` with fast projectiles when they are created and deleted almost instantly.

But why does it even happen if Replicate is removed before despawning ?

**Comment by cBournhonesque** [comment]
Yes the example seems incorrect, see this PR: https://github.com/cBournhonesque/lightyear/pull/808

1. `commands.entity(entity).remove::<server::Replicate>().despawn()` would still send a despawn message. It's best to use `despawn_without_replication()`
2. In general you always want to have the 'true' behaviour come from the server: so we want the server to replicate despawns, and the client to do 'predicted_despawns'. The server not sending a despawn is a mistake

**Comment by OlivierCoue** [comment]
Thanks for the clarification!

**Comment by OlivierCoue** [comment]
Not sure if its the correct place but to speak about it but, shouldn't we add ` || entity.get::<PreSpawnedPlayerObject>().is_some()` here:
https://github.com/cBournhonesque/lightyear/blob/86bad89d669ce5313c8beb7e566440ce9006d919/lightyear/src/client/prediction/despawn.rs#L43

Because if a prespawned entity get despawned with prediction_despawn before behind actually predicted (before the server send its owned entity with the PreSpawnedPlayerObject) then it falls in the case `error!("This command should only be called for predicted entities!");`

**Comment by cBournhonesque** [comment]
I think you might be right, we do want to treat `PreSpawnedPlayerObject` the same as a normal predicted entity before it's been matched with a server entity.

I also noticed we are not inserting any PredictionHistory on it, we only add the PredictionHistory at matching time, which might explain some of the rollbacks

**Comment by cBournhonesque** [comment]
I think I found the fix for the pre-spawned rollbacks: https://github.com/cBournhonesque/lightyear/pull/815

I tried running the spaceships and i get no rollbacks when spawning bullets. (there are still some rollbacks for avian collisions but that's an independent issue)
